contracts/pol/CarbonPOL.sol

```diff
@@ -0,0 +1,303 @@
+        if (currentPrice.ethAmount == 0) {
+            revert InvalidPrice();
+        }
+        // multiply the token amount by the eth amount / total eth amount ratio to get the actual tokens received
+        tokenAmount = uint128(MathEx.mulDivF(currentPrice.tokenAmount, ethAmount, currentPrice.ethAmount));
+        // revert if not enough token balance
+        if (tokenAmount > uint128(token.balanceOf(address(this)))) {
+            revert InsufficientTokenBalance();
+        }
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function expectedTradeInput(Token token, uint128 tokenAmount) public view returns (uint128 ethAmount) {
+        // return 0 if trading is not enabled for token
+        if (!_tradingEnabled(token)) {
+            return 0;
+        }
+        // revert if not enough token balance for trade
+        if (tokenAmount > token.balanceOf(address(this))) {
+            revert InsufficientTokenBalance();
+        }
+        Price memory currentPrice = tokenPrice(token);
+        // multiply the eth amount by the token amount / total token amount ratio to get the actual eth to send
+        return uint128(MathEx.mulDivF(currentPrice.ethAmount, tokenAmount, currentPrice.tokenAmount));
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function tokenPrice(Token token) public view returns (Price memory price) {
+        // return 0 if trading hasn't been enabled for a token
+        if (!_tradingEnabled(token)) {
+            return Price({ tokenAmount: 0, ethAmount: 0 });
+        }
+        // get initial price as set by enableTrading
+        price = _initialPrice[token];
+        // calculate the actual price by multiplying the eth amount by the factor
+        price.ethAmount *= _marketPriceMultiply;
+        // get time elapsed since trading was enabled
+        uint32 timeElapsed = uint32(block.timestamp) - _tradingStartTimes[token];
+        // get the current price by adjusting the eth amount with the exp decay formula
+        price.ethAmount = uint128(ExpDecayMath.calcExpDecay(price.ethAmount, timeElapsed, _priceDecayHalfLife));
+        // return the price
+        return price;
+    }
+
+    /**
+     * @inheritdoc ICarbonPOL
+     */
+    function trade(
+        Token token,
+        uint128 amount
+    ) external payable nonReentrant validToken(token) greaterThanZero(amount) {
+        uint128 ethRequired = expectedTradeInput(token, amount);
+        // revert if trade requires 0 eth
+        if (ethRequired == 0) {
+            revert InvalidTrade();
+        }
+        // check enough eth has been sent for the trade
+        if (msg.value < ethRequired) {
+            revert InsufficientNativeTokenSent();
+        }
+        // transfer the tokens to caller
+        token.safeTransfer(msg.sender, amount);
+
+        // refund any excess eth to caller
+        if (msg.value > ethRequired) {
+            payable(msg.sender).sendValue(msg.value - ethRequired);
```

Comment:
Super-duper nitpick: the subtraction can be `unchecked`

Reply:
Prefer to leave it as regular subtraction - doesn't save that much gas and makes the code less readable.

Reply:
👍